proxy/stats_handler.go

```diff
@@ -0,0 +1,215 @@
+package proxy
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	jwt "github.com/dgrijalva/jwt-go"
+	"github.com/gorilla/websocket"
+
+	"github.com/rancherio/websocket-proxy/common"
+)
+
+type StatsHandler struct {
+	backend         backendProxy
+	parsedPublicKey interface{}
+}
+
+func (h *StatsHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	multiHost := false
+
+	if strings.HasSuffix(req.URL.Path, "project/") || strings.HasSuffix(req.URL.Path, "service/") {
+		multiHost = true
+	}
+
+	hostKeys, urls, authed := h.auth(req, multiHost)
+	if !authed {
+		return
+	}
+
+	upgrader := websocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool { return true },
+	}
+	ws, err := upgrader.Upgrade(rw, req, nil)
+	if err != nil {
+		http.Error(rw, "Failed to upgrade connection.", 500)
+		return
+	}
+
+	msgKeys := make([]string, len(hostKeys))
+	respChannels := make([]<-chan common.Message, len(hostKeys))
+
+	var mutex sync.Mutex
+	var countMutex sync.Mutex
```

Comment:
I don't think you need these mutexes. The resources/variables they are protecting are only used in a single goroutine.
If I'm missing something, maybe you can explain it to me over slack.